gnuradio-runtime/lib/buffer_single_mapped.cc

```diff
@@ -0,0 +1,403 @@
+{
+    // Maybe adjust read pointers from min read index?
+    // This would mean that *all* readers must be > (passed) the write index
+    if (((d_bufsize - read_index) < (uint32_t)items_required) &&
+        (d_write_index < read_index)) {
+
+        // Find reader with the smallest read index that is greater than the
+        // write index
+        uint32_t min_reader_index = std::numeric_limits<uint32_t>::max();
+        uint32_t min_read_idx = std::numeric_limits<uint32_t>::max();
+        for (size_t idx = 0; idx < d_readers.size(); ++idx) {
+            if (d_readers[idx]->d_read_index > d_write_index) {
+                // Record index of reader with minimum read-index
+                if (d_readers[idx]->d_read_index < min_read_idx) {
+                    min_read_idx = d_readers[idx]->d_read_index;
+                    min_reader_index = idx;
+                }
+            }
+        }
+
+        // Note items_avail might be zero, that's okay.
+        items_avail += read_index - min_read_idx;
+        int gap = min_read_idx - d_write_index;
+        if (items_avail > gap) {
+            return false;
+        }
+
+#ifdef BUFFER_DEBUG
+        std::ostringstream msg;
+        msg << "[" << this << ";" << this << "] "
+            << "input_blocked_callback() WR_idx: " << d_write_index
+            << " -- WR items: " << nitems_written() << " -- BUFSIZE: " << d_bufsize
+            << " -- RD_idx: " << min_read_idx;
+        for (size_t idx = 0; idx < d_readers.size(); ++idx) {
+            if (idx != min_reader_index) {
+                msg << " -- OTHER_RDR: " << d_readers[idx]->d_read_index;
+            }
+        }
+
+        msg << " -- GAP: " << gap << " -- items_required: " << items_required
+            << " -- items_avail: " << items_avail;
+        GR_LOG_DEBUG(d_logger, msg.str());
+#endif
+
+        // Shift existing data down to make room for blocked data at end of buffer
+        uint32_t move_data_size = d_write_index * d_sizeof_item;
+        char* dest = buffer_ptr + (items_avail * d_sizeof_item);
+        memmove_func(dest, buffer_ptr, move_data_size);
+
+        // Next copy the data from the end of the buffer back to the beginning
+        uint32_t avail_data_size = items_avail * d_sizeof_item;
+        char* src = buffer_ptr + (min_read_idx * d_sizeof_item);
+        memcpy_func(buffer_ptr, src, avail_data_size);
+
+        // Now adjust write pointer
+        d_write_index += items_avail;
+
+        // Finally adjust all reader pointers
+        for (size_t idx = 0; idx < d_readers.size(); ++idx) {
+            if (idx == min_reader_index) {
+                d_readers[idx]->d_read_index = 0;
+            } else {
+                d_readers[idx]->d_read_index += items_avail;
+                d_readers[idx]->d_read_index %= d_bufsize;
+            }
+        }
+
+        return true;
+    }
+
+    return false;
+}
+
+bool buffer_single_mapped::output_blocked_callback_logic(int output_multiple,
```

Comment:
Both input and output blocked callback logic should be documented in code to describe how they work and when they are used

Reply:
Agreed, I will add the missing documentation.